lib/state/domain/tags.js

```diff
@@ -28,54 +28,47 @@ export const createTag = ({ name }) => {
   tagBucket().update(tagId, { name });
 };
 
-export const reorderTags = ({ tags }) => () => {
-  for (let i = 0; i < tags.length; i++) {
-    let tag = tags[i];
-    tag.data.index = i;
-    tagBucket().update(tag.id, tag.data);
-  }
-};
+export const reorderTags = ({ tags }) => () =>
+  tags.forEach((tag, index) =>
+    tagBucket().update(tag.id, { ...tag.data, index })
+  );
+
+export const renameTag = ({ tag, name: newName }) => (dispatch, getState) => {
+  const { notes } = getState().appState;
+  const tagName = tag.data.name;
 
-export const renameTag = ({ tag, name }) => (dispatch, getState) => {
-  let oldTagName = tag.data.name;
-  if (oldTagName === name) {
+  if (tagName === newName) {
     return;
   }
 
-  let { notes } = getState().appState;
-
-  tag.data.name = name;
-
-  tagBucket().update(tag.id, tag.data);
+  tagBucket().update(tag.id, { ...tag.data, name: newName });
 
-  // Replace tags in notes
-  for (let i = 0; i < notes.length; i++) {
-    let note = notes[i];
-    let noteTags = note.data.tags || [];
-    let tagIndex = noteTags.indexOf(oldTagName);
-
-    if (tagIndex !== -1) {
-      noteTags.splice(tagIndex, 1, name);
-      note.data.tags = noteTags.filter(noteTag => noteTag !== oldTagName);
-      noteBucket().update(note.id, note.data);
-    }
-  }
+  notes
```

Comment:
Why doesn't this happen in Simperium and then update the client?

Reply:
this is a point where Simperium and Simplenote diverge. Simperium isn't aware of `notes` or `tags` - it only knows about `buckets` and `entities`, so as an application-specific semantic we need to enforce it here.

it's also possible to add supervisory _watchers_ which are app-level admins that can watch the changes as they come in from clients. these watchers could look for tag changes, make additional note updates, and then have those changes distributed, but doing so from a practical standpoint would probably be significantly more complicated than simply repeating the logic here in the clients when it happens.

this is a subjective call and most reasoning probably goes along the lines of "because the data model is so small and contained, duplicating the logic in each client is the most straightforward way to enforce the behaviors"

good question!